---
title: "Nina Diff Expr Analyses"
output: html_document
---
---
title: "Deseq2 RNAseq d05,07 human ES diff to insulin+ cells analysis testing diff between experiments"
author: "Kristian Honnens de Lichtenberg"
date: "`r Sys.Date()`"
---


```{r setupKNITR, include=FALSE}
knitr::opts_chunk$set(cache=T)
```


```{r libraries, include=F, echo=FALSE}
library(knitr)
opts_chunk$set(root.dir ="/Users/Kristians/Dropbox/Kristian<->Palle/Nina_hES_d5_d7")
source("http://bioconductor.org/biocLite.R")

#biocLite(c( "DESeq2","ggplot2", "readxl", "EDASeq", "org.Hs.eg.db"), suppressUpdates=F, suppressAutoUpdate=F, ask=F)
library(ggplot2)
library(DESeq2)
library(superheat)
library(RColorBrewer)

library(readxl)
library(magrittr)
library(dplyr)
library(openxlsx)
```

Read DDS object from HUES4 & H1 project, remvoe  H1 clel line columns.
```{r read DDS}

dds<-readRDS("dds_HUES4_H1.RDS")

dds <- DESeqDataSetFromMatrix(countData = counts(dds),colData=colData(dds), design= ~Treatment) 


```

```{r vsdPCA d5}
vsd <- vst(dds[,dds$Diff_day=="D5"], blind=FALSE)

pcaData <- plotPCA(vsd, intgroup=colnames(colData(dds)), returnData=TRUE, ntop=5000)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Treatment, shape=Cell_line)) +
  geom_point(size=3) + scale_color_brewer(palette = "Set2") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()+ggtitle("PCA: Before Combat")
#ggsave(filename = "PCA_treatmentCellLine.pdf")

#treatment diffday
ggplot(pcaData, aes(PC1, PC2, color=Treatment, shape=Exp_batch)) +
  geom_point(size=3) + scale_color_brewer(palette = "Set2") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()+ggtitle("PCA: Before Combat")
#ggsave(filename = "PCA_treatmentDiffDay_D5_beforeCombat.pdf")

```

```{r vsdPCA d7}
vsd <- vst(dds[,dds$Diff_day=="D7"], blind=FALSE)

pcaData <- plotPCA(vsd, intgroup=colnames(colData(dds)), returnData=TRUE, ntop=5000)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=Treatment, shape=Exp_batch)) +
  geom_point(size=3) + scale_color_brewer(palette = "Set2") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
#ggsave(filename = "PCA_treatmentCellLine.pdf")

#treatment diffday
ggplot(pcaData, aes(PC1, PC2, color=Treatment, shape=Exp_batch)) +
  geom_point(size=3) + scale_color_brewer(palette = "Set2") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()+ggtitle("PCA: Before Combat")
ggsave(filename = "PCA_treatmentDiffDay_D7_beforeCombat.pdf")

```


```{r SVA-Seq d5}
#biocLite("sva")
library("sva")
dds_d5<-dds[,which(colData(dds)$Diff_day=="D5")]

dat  <- counts(dds_d5, normalized = TRUE)
idx  <- rowMeans(dat) > 1
dat  <- dat[idx, ]
mod  <- model.matrix(~Treatment, colData(dds_d5))
mod0 <- model.matrix(~   1, colData(dds_d5))
svseq <- svaseq(dat, mod, mod0, n.sv = 6)
svseq$sv

par(mfrow = c(2, 3), mar = c(3,5,3,1))
for (i in 1:6) {
  stripchart(svseq$sv[, i] ~ dds_d5$Exp_batch, vertical = TRUE, main = paste0("SV", i))
  abline(h = 0)
 }

#based on above, used the first two surrogate variables.
dds_d5$SV1 <- svseq$sv[,1]
dds_d5$SV2 <- svseq$sv[,2]

design(dds_d5) <- ~ SV1+SV2 + Treatment

```




```{r D5 only}
#dds_d5<-dds[,which(colData(dds)$Diff_day=="D5")]
dds_d5<-DESeq(dds_d5)
(treatmentList<-resultsNames(dds_d5)[-c(1:3)])

#up and down, LogFC=0.58 threshold:
treatmentResult<-lapply(treatmentList, function(x) results(dds_d5, name=x, lfcThreshold = 0.58))
names(treatmentResult)<-c("D5_IWP_vs_BM"  ,   "D5_IWP.TGF_vs_BM" ,"D5_TGF_vs_BM" )
lapply(treatmentResult[1:3], function(x) summary(x))
treatmentD5<-lapply(treatmentResult, function(x) data.frame(x))


write.xlsx(treatmentD5[1],paste0("Results_",names(treatmentD5[1]),".xlsx"),sheetName=paste0(names(treatmentD5[1])), col.names=TRUE, row.names=TRUE, append=FALSE)
write.xlsx(treatmentD5[2],paste0("Results_",names(treatmentD5[2]),".xlsx"),sheetName=paste0(names(treatmentD5[2])), col.names=TRUE, row.names=TRUE, append=FALSE)
write.xlsx(treatmentD5[3],paste0("Results_",names(treatmentD5[3]),".xlsx"),sheetName=paste0(names(treatmentD5[3])), col.names=TRUE, row.names=TRUE, append=FALSE)
```


```{r D7 only}
dds_D7<-dds[,which(colData(dds)$Diff_day=="D7")]

dat  <- counts(dds_D7, normalized = TRUE)
idx  <- rowMeans(dat) > 1
dat  <- dat[idx, ]
mod  <- model.matrix(~Treatment, colData(dds_D7))
mod0 <- model.matrix(~   1, colData(dds_D7))
svseq <- svaseq(dat, mod, mod0, n.sv = 6)
svseq$sv

par(mfrow = c(2, 3), mar = c(3,5,3,1))
for (i in 1:6) {
  stripchart(svseq$sv[, i] ~ dds_D7$Exp_batch, vertical = TRUE, main = paste0("SV", i))
  abline(h = 0)
 }

#based on above, used the first two surrogate variables.
dds_D7$SV1 <- svseq$sv[,1]
dds_D7$SV2 <- svseq$sv[,2]

design(dds_D7) <- ~ SV1+SV2 + Treatment




dds_D7<-DESeq(dds_D7)
resultsNames(dds_D7)
(treatmentList<-resultsNames(dds_D7)[-c(1:3)])

#up and down, LogFC=0.58 threshold:
treatmentResult<-lapply(treatmentList, function(x) results(dds_D7, name=x, lfcThreshold = 0.58))
names(treatmentResult)<-c("D7_IWP_vs_BM"  ,   "D7_IWP.TGF_vs_BM" ,"D7_TGF_vs_BM" )
lapply(treatmentResult[1:3], function(x) summary(x))
treatmentD7<-lapply(treatmentResult, function(x) data.frame(x))


write.xlsx(treatmentD7[1],paste0("Results_",names(treatmentD7[1]),".xlsx"),sheetName=paste0(names(treatmentD7[1])), col.names=TRUE, row.names=TRUE, append=FALSE)
write.xlsx(treatmentD7[2],paste0("Results_",names(treatmentD7[2]),".xlsx"),sheetName=paste0(names(treatmentD7[2])), col.names=TRUE, row.names=TRUE, append=FALSE)
write.xlsx(treatmentD7[3],paste0("Results_",names(treatmentD7[3]),".xlsx"),sheetName=paste0(names(treatmentD7[3])), col.names=TRUE, row.names=TRUE, append=FALSE)
```



```{r collect sign genes for heatmap d5}
library(superheat)


all_genes<-lapply(treatmentD5, function(x) rownames(x)[abs(x$log2FoldChange)>1&x$padj<0.01])
all_unlist<-unlist(all_genes)
all_undup<-unique(all_unlist)
length(all_undup)

count_matrix<-counts(dds_d5)

colData(dds_d5)$SampleName<-interaction(colData(dds_d5)$Treatment, colData(dds_d5)$Exp_batch)

count_D5DE<-count_matrix[which(rownames(count_matrix) %in% all_undup),]

count_D5DE<-count_D5DE[!is.na(count_D5DE[,1]),]


dat_scaled<- t(scale(t(count_D5DE)))
colnames(dat_scaled)<-dds_d5$SampleName


png("d5_heatmap_SampleName.png", height = 600, width = 500)
  set.seed(108); 
  superheat(dat_scaled,
            #membership.cols = dds_d5$Treatment,
    order.cols = order(as.character(dds_d5$SampleName)),
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=5,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 5 \n Differentially Expressed Genes"),
   bottom.label.text.angle = 90, 
     bottom.label.size = 0.3,
      left.label.size = 0.1,
     bottom.label.text.size = 4,
     bottom.label.text.alignment = "center")
dev.off()

png("d5_heatmap_Treatment.png", height = 600, width = 500)
set.seed(108); 
  superheat(dat_scaled, membership.cols = dds_d5$Treatment,
            order.cols = order(as.character(dds_d5$SampleName)),
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=5,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 5 \n Differentially Expressed Genes"),
     bottom.label.text.angle = 0, 
     bottom.label.size = 0.1,
     bottom.label.text.size = 5,
     bottom.label.text.alignment = c("center"))
dev.off()
     
png("d5_heatmap_TreatmentSmooth.png", height = 600, width = 500)
set.seed(108); 
  superheat(dat_scaled, membership.cols = dds_d5$Treatment,
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=5,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 5 \n Differentially Expressed Genes"),
     bottom.label.text.angle = 0, 
     bottom.label.size = 0.1,
     bottom.label.text.size = 5,
     bottom.label.text.alignment = c("center"), smooth.heat = T)
dev.off()
```


```{r collect sign genes for heatmap d7}

all_genes<-lapply(treatmentD7, function(x) rownames(x)[abs(x$log2FoldChange)>1&x$padj<0.05])
all_unlist<-unlist(all_genes)
all_undup<-unique(all_unlist)
length(all_undup)

count_matrix<-counts(dds_D7)

colData(dds_D7)$SampleName<-interaction(colData(dds_D7)$Treatment, colData(dds_D7)$Exp_batch)

count_D7DE<-count_matrix[which(rownames(count_matrix) %in% all_undup),]

count_D7DE<-count_D7DE[!is.na(count_D7DE[,1]),]


dat_scaled<- t(scale(t(count_D7DE)))
colnames(dat_scaled)<-dds_D7$SampleName

png("d7_heatmap_SampleName.png", height = 600, width = 500)
  set.seed(108); 
  superheat(dat_scaled, 
            #membership.cols = dds_D7$Treatment,
            order.cols = order(as.character(dds_D7$SampleName)),
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=5,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 7 \n Differentially Expressed Genes"),
     bottom.label.text.angle = 90, 
     bottom.label.size = 0.3,
      left.label.size = 0.1,
     bottom.label.text.size = 4,
     bottom.label.text.alignment = "center", column.title = "Variables")
dev.off()

png("d7_heatmap_Treatment.png", height = 600, width = 500)
set.seed(108); 
  superheat(dat_scaled, 
            membership.cols = dds_D7$Treatment,
                        order.cols = order(as.character(dds_D7$SampleName)),
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=5,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 7 \n Differentially Expressed Genes"),
     bottom.label.text.angle = 0, 
     bottom.label.size = 0.1,
     bottom.label.text.size = 5,
     bottom.label.text.alignment = c("center"))
dev.off()
     
png("d7_heatmap_TreatmentSmooth.png", height = 600, width = 500)
set.seed(108); 
  superheat(dat_scaled, membership.cols = dds_D7$Treatment,
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=5,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 7 \n Differentially Expressed Genes"),
     bottom.label.text.angle = 0, 
     bottom.label.size = 0.1,
     bottom.label.text.size = 5,
     bottom.label.text.alignment = c("center"), smooth.heat = T)
dev.off()


  superheat(dat_scaled, membership.cols = dds_D7$Treatment,
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=4,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 7 \n Differentially Expressed Genes"),
     bottom.label.text.angle = 0, 
     bottom.label.size = 0.1,
     bottom.label.text.size = 5,
     bottom.label.text.alignment = c("center"), smooth.heat = T)
  
  
    superheat(dat_scaled, membership.cols = dds_D7$Treatment,
     heat.pal = brewer.pal(5,"YlOrRd"), 
     pretty.order.cols = T,
     scale=F,
     n.clusters.rows=3,
     dist.method=c("manhattan"),
     clustering.method=c("kmeans"),
     title=c("Heatmap of Day 7 \n Differentially Expressed Genes"),
     bottom.label.text.angle = 0, 
     bottom.label.size = 0.1,
     bottom.label.text.size = 5,
     bottom.label.text.alignment = c("center"), smooth.heat = T)

```

```{r boxplots d5, eval=F}
library(reshape2); library(dplyr)
geneexpression<-fpkm(dds_d5) #do FPKM


#adjust to NINAs dataset:
identical(colnames(d10d13_fpkm_symbol),sampleTable$sampleName)
colnames(geneexpression)<-sampleTable$grpday
geneexpressionmelted<-melt(geneexpression)

geneexpressionmelted$Var2<-factor(geneexpressionmelted$Var2)
geneexpressionmelted$Var2= factor(geneexpressionmelted$Var2,levels(geneexpressionmelted$Var2)[c(5,7,6,8,3,1,2,4)])

colnames(geneexpressionmelted)<-c("Gene", "Genotype", "FPKM")
geneexpressionmelted$Gene<-as.character(geneexpressionmelted$Gene)

geneexpressionmelted$Genotype<-dplyr::recode(geneexpressionmelted$Genotype,Day10.HES1wt="Day10.Wildtype",Day13.HES1wt="Day13.Wildtype" ,.default = levels(geneexpressionmelted$Genotype))






##test genelist

genenamelist<-c("ASCL1","NEUROG3", "HES1", "ATOH1" )

#plot boxplot function:
plotBoxPlot<-function(genenamelist){
  for(i in genenamelist) { 
p<-ggplot(geneexpressionmelted[geneexpressionmelted$Gene==paste0(i),], aes(Treatment,FPKM, fill = Treatment)) +geom_boxplot(show.legend=F)+theme(axis.text=element_text(size=17), plot.title=element_text(size=25,face="italic"), axis.title.y = element_text(size=17))+theme(axis.text.x = element_text(vjust = 1, hjust=1, angle=45))
print(p+ ggtitle(paste0(i)))
ggsave(paste0(i, ".png"), plot = last_plot(), device = NULL, path = "./boxplots_d5")
}
}

#rewrite for NINAs project:
listofobjects<-list(Day10.H1N3_dKO_vs_Day10.HES1KO@listData,Day10.HES1KO_vs_Day10.HES1wt@listData,Day10.NGN3KO_vs_Day10.HES1wt@listData,Day13.H1N3_dKO_vs_Day10.H1N3_dKO@listData,Day13.H1N3_dKO_vs_Day13.NGN3KO@listData,Day13.HES1KO_vs_Day10.HES1KO@listData,Day13.HES1wt_vs_Day10.HES1wt@listData,Day13.NGN3KO_vs_Day10.NGN3KO@listData,grpday_Day10.HES1KO_vs_Day10.HES1wt@listData)
names(listofobjects)<-ls(pattern="_vs_")
#check duplication by unique()

#plotboxplots:

```


```{r}
sessionInfo()
```

